src/execution_plan/execution_plan.c

```diff
@@ -330,39 +330,53 @@ static void _ExecutionPlan_PlaceApplyOps(ExecutionPlan *plan) {
 	}
 	array_free(filter_ops);
 }
+
+void ExecutionPlan_RePositionFilterOp(ExecutionPlan *plan, OpBase *search_root,
+									  const OpBase *recurse_limit, OpBase *filter) {
+	assert(filter->type == OPType_FILTER);
+	rax *references = FilterTree_CollectModified(((OpFilter *)filter)->filterTree);
+	OpBase *op;
+	if(raxSize(references) > 0) {
+		/* Scan execution plan, locate the earliest position where all
+		 * references been resolved. */
+		op = ExecutionPlan_LocateReferences(search_root, recurse_limit, references);
+	} else {
+		/* The filter tree does not contain references, like:
+		 * WHERE 1=1
+		 * TODO This logic is inadequate. For now, we'll place the op
+		 * directly below the first projection (hopefully there is one!). */
+		op = plan->root;
+		while(op->childCount > 0 && op->type != OPType_PROJECT && op->type != OPType_AGGREGATE) {
+			op = op->children[0];
+		}
+	}
+	assert(op);
+	// In case this is a pre-existing filter (this function is not called out from ExecutionPlan_PlaceFilterOps)
+	if(filter->childCount > 0) {
+		// If the located op is not the filter child, re position the filter.
+		if(op != filter->children[0]) {
+			ExecutionPlan_RemoveOp(plan, (OpBase *)filter);
+			ExecutionPlan_PushBelow(op, (OpBase *)filter);
+		}
+	} else {
+		// This is a new filter.
+		ExecutionPlan_PushBelow(op, (OpBase *)filter);
+	}
```

Comment:
Why the distinction?
Why should we care about the child of filter?

I think that at the very beginning of the function we should disconnect filter from its children.

Reply:
you don't want to insert it in the exact same location